README
= midilib

midilib is a pure Ruby MIDI library useful for reading and writing standard
MIDI files and manipulating MIDI event data. Classes include MIDI::Sequence,
MIDI::Track, MIDI::Event, and MIDI::IO::MIDIFile and its subclasses
MIDI::IO::SeqReader and MIDI::IO::SeqWriter.

The Web site of midilib is (http://midilib.rubyforge.org). The RubyForge
project page is http://rubyforge.org/projects/midilib, where the latest
version of midilib may be downloaded. midilib is also available as a RubyGem.


== Dependencies

midilib does not require any other packages. The test suite in the tests
directory requires the testing framework TestUnit, which comes with Ruby 1.8
and later and can also be found in the Ruby Application Archive
(http://raa.ruby-lang.org).

To rebuild the gem or RDocs or run the tests easily, you can use the Rakefile
which requires Rake (http://rake.rubyforge.org).


== Installation

=== RubyGems Installation

To install midilib as a gem, type

    % gem install midilib
or
    % gem update midilib

if you already have a previous version. You may need root privileges to
install or update the gem.

=== Manual Installation

After downloading and expanding the archive, you can install midilib with the
command

    % ruby install.rb
(or)
    % ruby install.rb --install-dir=my_directory

You may need root privileges to install.


== Testing

  % rake test

runs all of the tests in the test directory.


== Overview

MIDI file IO only understands MIDI file format 1, where a sequence is made up
of multiple tracks. It doesn't yet understand format 0 (a single track
containing all events) or format 2 (a collection of format 0 files in one
file).

=== MIDI::Sequence

A sequence contains a collection of tracks and global information like the
sequence's pulses per quarter note (ppqn) and time signature.

The first track in a sequence is special; it holds meta-events like tempo and
sequence name. Don't put any notes in this track.

MIDI::Sequence also contains some convenience methods that let you set and
retrieve the sequence's name, the time signature, and to retrieve the first
tempo event's beats-per-minute value.

Normally instances of MIDI::IO::SeqReader and MIDI::IO::SeqWriter are used
when a sequence reads itself from or writes itself to a MIDI file. You can
change that by setting a sequence's reader_class or writer_class attributes.
Instances of the classes contained in those attributes are created and used
whenever the sequence reads or writes itself.

=== MIDI::Track

A track contains an array of events.

When you modify the +events+ array, make sure to call recalc_times so each
event gets its +time_from_start+ recalculated. You don't have to do that
after every event you add; just remember to do so before using the track in a
way that expects the list of events to be ordered correctly.

A Track also holds a bit mask that specifies the channels used by the track.
This bit mask is set when the track is read from the MIDI file by a SeqReader
but is _not_ kept up to date by any other methods. Specifically, if you add
events to a track at any other time, the bit mask will not be updated.

=== MIDI::Measure

This class contains information about a measure from the sequence. Measure
data is based on the time signature information from the sequence and is not
stored in the sequence itself.

=== MIDI::Measures

The class MIDI::Sequence method get_measures returns a MIDI::Measures object.
MIDI::Measures is a subclass of Array. It is a specialized container for
MIDI::Measure objects, which can be use to map event times to measure numbers.
Please note that this object has to be remade when events are deleted/added in
the sequence.

MIDI::Measure and MIDI::Measures are brought to us by Jari Williamsson
<jari.williamsson@mailbox.swipnet.se>, who also contributed some improvements
to the MIDI::Event and MIDI::Track classes.

=== MIDI::Event

Each event holds not only its delta time but also its time from the start of
the track. The track is responsible for recalculating its events' start times.
You can call MIDI::Track#recalc_times to do so.

Events have a number of boolean methods that identify their types, like
channel?, note?, note_on?, note_off?, meta?, system?, realtime?, and
program_change?. Events know how to print themselves. By default, all numbers
are printed as hexidecimal and channel numbers are printed from 0-15. Setting
print_decimal_numbers to true will output decimal numbers and setting
print_channel_numbers_from_one will output channel numbers from 1-16.


Subclasses of MIDI::Event implement the various MIDI messages such as note on
and off, controller values, system exclusive data, and realtime bytes.

MIDI::Realtime events have delta values and start times, just like all the
other midilib event types do. (MIDI real time status bytes don't have delta
times, but this way we can record when in a track the realtime byte was
received and should be sent. This is useful for start/continue/stop events
that control other devices, for example.) Note that when a MIDI::Realtime
event is written out to a MIDI file, the delta time is not written.


== How To Use

The following examples show you how to use midilib to read, write, and
manipulate MIDI files and modify track events. See also the files in the
examples directory, which are described below.


=== Reading a MIDI File

To read a MIDI file, create a MIDI::Sequence object and call its #read method,
passing in an IO object.

The #read method takes an optional block. If present, the block is called once
after each track has finished being read. Each time, it is passed the total
number of tracks and the number of the current track that has just been read.
This is useful for notifying the user of progress, for example by updating a
GUI progress bar.

 require 'midi/io/seqreader'

 # Create a new, empty sequence.
 seq = MIDI::Sequence.new()

 # Read the contents of a MIDI file into the sequence.
 File.open('my_midi_file.mid', 'rb') { | file |
     seq.read(file) { | num_tracks, i |
         # Print something when each track is read.
         puts "read track #{i} of #{num_tracks}"
     }
 }


=== Writing a MIDI File

To write a MIDI file, call the write method, passing in an IO object.


 require 'midi/io/seqwriter'

 # Start with a sequence that has something worth saving.
 seq = read_or_create_seq_we_care_not_how()

 # Write the sequence to a MIDI file.
 File.open('my_output_file.mid', 'wb') { | file | seq.write(file) }


=== Editing a MIDI File

Combining the last two examples, here is a script that reads a MIDI file,
transposes some events, and writes the sequence out to a different file. This
is a useful template for programatically manipulating MIDI data.


This code transposes all of the note events (note on, note off, and poly
pressure) on channel 5 down one octave. It's easy to find events that need to
be transposed: the method #note? returns true.

==== Transposing One Channel

 require 'midi/io/seqreader'
 require 'midi/io/seqwriter'

 # Create a new, empty sequence.
 seq = MIDI::Sequence.new()

 # Read the contents of a MIDI file into the sequence.
 File.open('my_input_file.mid', 'rb') { | file |
     seq.read(file) { | num_tracks, i |
         # Print something when each track is read.
         puts "read track #{i} of #{num_tracks}"
     }
 }

 # Iterate over every event in every track.
 seq.each { | track |
     track.each { | event |
         # If the event is a note event (note on, note off, or poly
         # pressure) and it is on MIDI channel 5 (channels start at
         # 0, so we use 4), then transpose the event down one octave.
         if event.note? && event.channel == 4
             event.note -= 12
         end
     }
 }

 # Write the sequence to a MIDI file.
 File.open('my_output_file.mid', 'wb') { | file | seq.write(file) }


=== Manipulating tracks

If you modify a track's list of events directly, don't forget to call
MIDI::Track#recalc_times when you are done.

 track.events[42, 1] = array_of_events
 track.events << an_event
 track.merge(array_of_events)
 track.recalc_times

=== Calculating delta times

A few methods in MIDI::Sequence make it easier to calculate the delta times
that represent note lengths. MIDI::Sequence#length_to_delta takes a note
length (a multiple of a quarter note) and returns the delta time given the
sequence's current ppqn (pulses per quarter note) setting. 1 is a quarter
note, 1.0/32.0 is a 32nd note (use floating-point numbers to avoid integer
rounding), 1.5 is a dotted quarter, etc. See the documentation for that method
for more information.

MIDI::Sequence#note_to_length takes a note name and returns a length value
(again, as a multiple of a quarter note). Legal note names are those found in
MIDI::Sequence::NOTE_TO_LENGTH, and may begin with "dotted" and/or end with
"triplet". For example, "whole", "sixteenth", "32nd", "quarter triplet",
"dotted 16th", and "dotted 8th triplet" are all legal note names.

Finally, MIDI::Sequence#note_to_delta takes a note name and returns a delta
time. It does this by calling note_to_length, then passing the result to
length_to_delta.


=== Example Scripts

Here are short descriptions of each of the examples found in the examples
directory.

* examples/from_scratch.rb shows you how to create a new sequence from scratch
  and save it to a MIDI file. It creates a file called 'from_scratch.mid'.

* examples/seq2text.rb dumps a MIDI file as text. It reads in a sequence and
  uses the to_s method of each event.

* examples/reader2text.rb dumps a MIDI file as text. It subclasses
  MIDI::SeqReader instead of creating a sequence containing tracks and events.

* examples/transpose.rb transposes all note events (note on, note off, poly
  pressure) on a specified channel by a specified amount.

* There is also one MIDI file, examples/NoFences.mid. It is a little pop ditty
  I wrote. The instruments in this file use General MIDI patch numbers and
  drum note assignments. Since I don't normally use GM patches, the sounds
  used here are at best approximations of the sounds I use.


== Resources

The Ruby Web site (http://www.ruby-lang.org/en/index.html) contains an
introduction to Ruby, the Ruby Application Archive (RAA) at
http://raa.ruby-lang.org, and pointers to more information.


<cite>Programming Ruby, The Pragmatic Programmer's Guide</cite>, by David
Thomas and Andrew Hunt, is a well-written and practical introduction to Ruby.
Its Web page at http://www.rubycentral.com/book also contains a wealth of Ruby
information. Though the first edition book is available online, I encourage
you to purchase a copy of the latest edition.

A description of the MIDI file format can be found in a few places such as
http://www.borg.com/~jglatt/tech/midifile.htm.

The MIDI message reference at http://www.io.com/~jimm/midi_ref.html describes
the format of MIDI commands.


= To Do

:include: TODO


= Support

* Visit the forums, bug list, and mailing list pages at
  http://rubyforge.org/projects/midilib

* Send email to Jim Menard at mailto:jimm@io.com

* Ask on the ruby-talk mailing list


= Administrivia

Author:: Jim Menard (mailto:jimm@io.com)
Copyright:: Copyright (c) 2003-2005 Jim Menard
License:: Distributed under the same license as Ruby.


== Copying

midilib is copyrighted free software by Jim Menard and is released under the
same license as Ruby. See the Ruby license at
http://www.ruby-lang.org/en/LICENSE.txt.

midilib may be freely copied in its entirety providing this notice, all
source code, all documentation, and all other files are included.

midilib is Copyright (c) 2003-2005 by Jim Menard.

The song "No Fences" contained in the MIDI file examples/NoFences.mid is
Copyright (c) 1992 by Jim Menard (jimm@io.com). It may be freely used for
non-commercial purposes as long as the author is given credit.


=== Recent Changes

==== New code repository

The midilib code is now hosted at Github (http://github.com/jimm/midilib).

==== Changes for 1.1.4:

* Fixed a bug in KeySig.data_as_bytes. Thanks to Noah Thorp for finding this
  and the bug fixed in 1.1.3.

==== Changes for 1.1.3:

* Fixed the way midilib detects the behavior of IO.getc.

==== Changes for 1.1.2:

* Define MIDI::IO::MIDIFile.getc differently for different Ruby versions,
  instead of checking for String.bytes every time we read a byte.

==== Changes for 1.1.1:

* Make MIDI::IO::MIDIFile.getc do the right thing for both Ruby 1.8 and 1.9.

==== Changes for 1.1.0:

* Added test/test.mid to list of files to be included when packaging midifile
  for distribution.

==== Changes for 1.0.0:

* Fixed the bug in Track#recalc_delta_from_times found by Christopher Rose.

==== Changes for 0.8.7:

* Fixed the misspelled POLY_PRESSURE constant, thanks to Mario Pehle.

==== Changes for 0.8.6:

* Added missing test/test.mid.

==== Changes for 0.8.5:

* Fixed bugs in MIDI::PitchBend reading and writing, thanks to Emanuel
  Borsboom.

* Fixed a bug in MIDI::Track#quantize.

* The argument to MIDI::Track#quantize has changed: it is now either a note
  name ("sixteenth", "32nd", "8th triplet") or a length (1 = quarter, 0.25 =
  sixteenth). This is a drastic change that will break all previous calls to
  quantize. However, since that method was broken already, I don't feel it's
  a burden to anybody to change the arguments.

==== Changes for 0.8.4:

* Realtime status bytes now set @is_realtime to true and return true when
  realtime? is called.

* All system common events now set @is_system to true and return true when
  system? is called, not just system exclusive events.

* Added examples/from_scratch.rb, which shows how to create a sequence
  manually.

* New MIDI::Sequence methods that turn note length names like "32nd", "dotted
  quarter", and "16th triplet" into delta times. See the docs below and
  MIDI::Sequence::length_to_delta, MIDI::Sequence::note_to_length, and
  MIDI::Sequence::note_to_delta.


==== Changes for 0.8.3:

* Added MIDI::NoteEvent.note_to_s, which returns note name as a string like
  "C4" or "F#6".

* Added new boolean attributes to MIDI::Event: @print_decimal_numbers and
  @print_note_names. These are used by all Event to_s methods. See
  examples/seq2text.rb for an example.

==== Changes for 0.8.2:

* Changed MIDI::MetaEvent.type to MIDI::MetaEvent.event_type to avoid
  runtime complaints about Object#type calls.
* Added 'b' binary flag to file open modes for Windows.
* Fixed $LOAD_PATH in example files.
* Fixed read and write block arguments.
* Fixed other example script bugs.

==== Changes for 0.8.1:

* Fixed track sorting.
* Fixed track's recalc_delta_from_times method.
* Fixed event quantization.
* More tests and documentation.


== Warranty

This software is provided "as is" and without any express or implied
warranties, including, without limitation, the implied warranties of
merchantability and fitness for a particular purpose.

--
# Local Variables:
# mode:rdoc
# End:
++


ChangeLog
This change log is no longer maintaned. For further change descriptions,
see the Git logs.

2007-12-11  Jim Menard  <jim@jimmenard.com>

	* lib/midilib/io/seqreader.rb: Fixed text method and added default
	implementation of key_signature.

2007-12-09  Jim Menard  <jim@jimmenard.com>

	* examples/measures_mbt.rb: Added.

2007-12-09  Jim Menard  <jim@jimmenard.com>

	* lib/midilib/event.rb: Added code that fixes bpm calculation
	bugs, and his TimeSig and KeySig classes.

	* lib/midilib/measure.rb: Added.

	* lib/midilib/sequence.rb: Fixed clock. Added get_measures method.

2007-12-09  Jim Menard  <jim@jimmenard.com>

	* lib/midilib/event.rb: added program_change? and
	print_channel_numbers_from_one to Event.

2007-03-17  Jim Menard  <jim@jimmenard.com>

	* Version 1.0.0 released. I will no longer be maintaining this
	change log; the Subversion comments should be sufficient.

2006-08-20  Jim Menard  <jim@jimmenard.com>

	* lib/midilib/event.rb (PolyPressure::initialize): Fixed the
	misspelled POLY_PRESSURE cosntant, thanks to Mario Pehle
	<rubyforge.org@errorinitus.de>.

2006-05-15  Jim Menard  <jimm@localhost.localdomain>

	* test/test.mid: created (copied examples/from_scratch.mid).

2005-03-21  Jim Menard  <jim@jimmenard.com>

	* Version 0.8.4 released.

	* lib/midilib/event.rb (Realtime::initialize): set @is_realtime
	to true, not false.
	(SystemCommon::initialize): moved @is_system = true to here
	(SystemExclusive::initialize): ...from here.

2005-03-20  Jim Menard  <jim@jimmenard.com>

	* lib/midilib/sequence.rb (Sequence::note_to_delta): created.
	(Sequence::note_to_length): created.
	(Sequence::length_to_delta): created.

	* examples/from_scratch.rb: created.

2004-07-16  Jim Menard  <jim@jimmenard.com>

	* Version 0.8.3 released.

2004-07-10  Jim Menard  <jim@jimmenard.com>

	* lib/midilib/event.rb (NoteEvent::note_to_s): created.
	(Event::number_to_s): created.
	(Event): added @print_note_names and @print_decimal_numbers
	attributes.
	(to_s all classes): use @print_note_names and @print_decimal_numbers

2004-06-30  Jim Menard  <jim@jimmenard.com>

	* Version 0.8.2 released.

	* lib/midilib/event.rb (MetaEvent): changed @type to @meta_type to
	avoid warnings like "Object#type is deprecated; use Object#class".

	* lib/midilib/track.rb (Track::name): use Event.meta_type (renamed
	from Event.type).
	(Track::name=): use Event.meta_type (renamed from Event.type).

	* test/event_equality.rb (MIDI::MetaEvent): use meta_type instead
	of type.

	* examples/transpose.rb: fixed $LOAD_PATH. Added 'b' to file open
	modes for Windows.

	* examples/strings.rb: fixed $LOAD_PATH. Fixed arguments passed to
	read block. Fixed code that looks for meta events. Added 'b' to
	file open modes for Windows.

	* examples/seq2text.rb: fixed $LOAD_PATH. Fixed arguments passed
	to read block. Added 'b' to file open modes for Windows.

	* examples/reader2text.rb: fixed $LOAD_PATH. Fixed TextTranslator
	superclass. Fixed arguments passed to read block. Added 'b' to
	file open modes for Windows.

	* lib/midilib/io/seqwriter.rb (SeqWriter::initialize): added block
	rdoc comment.
	(SeqWriter::write_to): added track to @update_block args.

	* lib/midilib/io/seqreader.rb (SeqReader::initialize): added block
	rdoc comment.

2004-06-27  Jim Menard  <jim@jimmenard.com>

	* Version 0.8.1 released.

	* test/test_event.rb: created.

	* lib/midilib/track.rb: more documentation.
	(Track::sort): sorts by events' time_from_start and modifies
	@events (which wasn't happening before; I forgot to assign the
	sorted results back to @events).
	(Track::recalc_delta_from_times): fixed.
	(Track::quantize): call recalc_delta_from_times.

	* test/test_track.rb (TrackTester::test_sort): created.
	(TrackTester::test_recalc_delta_from_times): created.

	* lib/midilib/sequence.rb: more documentation.

	* lib/midilib/consts.rb: hid some comments from RDoc.

	* lib/midilib/event.rb: more documentation.
	(Event::realtime): fix quantize_to so it changes the event's
	time_from_start instead of delta_time.

	* test/*.rb: removed redundant copyright and license notices.


lib/midilib/event.rb
require_relative 'consts'
require_relative 'utils'

module MIDI
  # The abstract superclass of all MIDI events.
  class Event
    # Modifying delta_time does not affect time_from_start. You need to call
    # the event's track's +recalc_time+ method.
    attr_accessor :delta_time
    # The start time of this event from the beginning of the track. This value
    # is held here but is maintained by the track.
    attr_accessor :time_from_start
    # The MIDI status byte. Never includes the channel, which is held
    # separately by MIDI::ChannelEvent.
    attr_reader :status

    # Determines if to_s outputs hex note numbers (false, the default) or
    # decimal note names (true).
    attr_accessor :print_note_names

    # Determines if to_s outputs numbers as hex (false, the default) or
    # decimal # (true). Delta times are always printed as decimal.
    attr_accessor :print_decimal_numbers

    # Determines if to_s outputs MIDI channel numbers from 1-16 instead
    # of the default 0-15.
    attr_accessor :print_channel_numbers_from_one

    def initialize(status = 0, delta_time = 0)
      @status = status
      @delta_time = delta_time
      @time_from_start = 0 # maintained by tracks
      @print_note_names = false
      @print_decimal_numbers = false
      @print_channel_numbers_from_one = false
    end
    protected :initialize

    # Returns the raw bytes that are written to a MIDI file or output to a
    # MIDI stream. In MIDI::EVENT this raises a "subclass responsibility"
    # exception.
    def data_as_bytes
      raise 'subclass responsibility'
    end

    # Quantize this event's time_from_start by moving it to the nearest
    # multiple of +boundary+. See MIDI::Track#quantize. *Note*: does not
    # modify the event's delta_time, though MIDI::Track#quantize calls
    # recalc_delta_from_times after it asks each event to quantize itself.
    def quantize_to(boundary)
      diff = @time_from_start % boundary
      @time_from_start -= diff
      @time_from_start += boundary if diff >= boundary / 2
    end

    # For sorting. Uses @time_from_start, which is maintained by this event's
    # track. I'm not sure this is necessary, since each track has to
    # maintain its events' time-from-start values anyway.
    def <=>(other)
      @time_from_start <=> other.time_from_start
    end

    # Returns +val+ as a decimal or hex string, depending upon the value of
    # @print_decimal_numbers.
    def number_to_s(val)
      @print_decimal_numbers ? val.to_s : ('%02x' % val)
    end

    # Returns +val+ as a decimal or hex string, depending upon the value of
    # @print_decimal_numbers.
    def channel_to_s(val)
      val += 1 if @print_channel_numbers_from_one
      number_to_s(val)
    end

    def to_s
      "#{@delta_time}: "
    end
  end

  # The abstract superclass of all channel events (events that have a MIDI
  # channel, like notes and program changes).
  class ChannelEvent < Event
    # MIDI channel, 0-15.
    attr_accessor :channel

    def initialize(status, channel, delta_time)
      super(status, delta_time)
      @channel = channel
    end
    protected :initialize

    def to_s
      super << "ch #{channel_to_s(@channel)} "
    end
  end

  # The abstract superclass of all note on, and note off, and polyphonic
  # pressure events.
  class NoteEvent < ChannelEvent
    attr_accessor :note, :velocity

    def initialize(status, channel, note, velocity, delta_time)
      super(status, channel, delta_time)
      @note = note
      @velocity = velocity
    end
    protected :initialize

    PITCHES = %w[C C# D D# E F F# G G# A A# B]

    # Returns note name as a pitch/octave string like "C4" or "F#6".
    def pch_oct(val = @note)
      pch = val % 12
      oct = (val / 12) - 1
      "#{PITCHES[pch]}#{oct}"
    end

    # If @print_note_names is true, returns pch_oct(val) else returns value
    # as a number using number_to_s.
    def note_to_s
      @print_note_names ? pch_oct(@note) : number_to_s(@note)
    end

    def data_as_bytes
      data = []
      data << (@status + @channel)
      data << @note
      data << @velocity
    end
  end

  class NoteOn < NoteEvent
    attr_accessor :off

    def initialize(channel = 0, note = 64, velocity = 64, delta_time = 0)
      super(NOTE_ON, channel, note, velocity, delta_time)
    end

    def to_s
      super <<
        "on #{note_to_s} #{number_to_s(@velocity)}"
    end
  end

  # Old class name for compatability
  NoteOnEvent = NoteOn

  class NoteOff < NoteEvent
    attr_accessor :on

    def initialize(channel = 0, note = 64, velocity = 64, delta_time = 0)
      super(NOTE_OFF, channel, note, velocity, delta_time)
    end

    def to_s
      super <<
        "off #{note_to_s} #{number_to_s(@velocity)}"
    end
  end

  # Old class name for compatability
  NoteOffEvent = NoteOff

  class PolyPressure < NoteEvent
    def initialize(channel = 0, note = 64, value = 0, delta_time = 0)
      super(POLY_PRESSURE, channel, note, value, delta_time)
    end

    def pressure
      @velocity
    end

    def pressure=(val)
      @velocity = val
    end

    def to_s
      super <<
        "poly press #{channel_to_s(@channel)} #{note_to_s} #{number_to_s(@velocity)}"
    end
  end

  class Controller < ChannelEvent
    attr_accessor :controller, :value

    def initialize(channel = 0, controller = 0, value = 0, delta_time = 0)
      super(CONTROLLER, channel, delta_time)
      @controller = controller
      @value = value
    end

    def data_as_bytes
      data = []
      data << (@status + @channel)
      data << @controller
      data << @value
    end

    def to_s
      super << "cntl #{number_to_s(@controller)} #{number_to_s(@value)}"
    end
  end

  class ProgramChange < ChannelEvent
    attr_accessor :program

    def initialize(channel = 0, program = 0, delta_time = 0)
      super(PROGRAM_CHANGE, channel, delta_time)
      @program = program
    end

    def data_as_bytes
      data = []
      data << (@status + @channel)
      data << @program
    end

    def to_s
      super << "prog #{number_to_s(@program)}"
    end
  end

  class ChannelPressure < ChannelEvent
    attr_accessor :pressure

    def initialize(channel = 0, pressure = 0, delta_time = 0)
      super(CHANNEL_PRESSURE, channel, delta_time)
      @pressure = pressure
    end

    def data_as_bytes
      data = []
      data << (@status + @channel)
      data << @pressure
    end

    def to_s
      super << "chan press #{number_to_s(@pressure)}"
    end
  end

  class PitchBend < ChannelEvent
    attr_accessor :value

    def initialize(channel = 0, value = 0, delta_time = 0)
      super(PITCH_BEND, channel, delta_time)
      @value = value
    end

    def data_as_bytes
      data = []
      data << (@status + @channel)
      data << (@value & 0x7f) # lsb
      data << ((@value >> 7) & 0x7f) # msb
    end

    def to_s
      super << "pb #{number_to_s(@value)}"
    end
  end

  class SystemCommon < Event
    def initialize(status, delta_time)
      super(status, delta_time)
    end
  end

  class SystemExclusive < SystemCommon
    attr_accessor :data

    def initialize(data, delta_time = 0)
      super(SYSEX, delta_time)
      @data = data
    end

    def data_as_bytes
      data = []
      data << @status
      data << Utils.as_var_len(@data.length)
      data << @data
      data << EOX
      data.flatten
    end

    def to_s
      super << 'sys ex'
    end
  end

  class SongPointer < SystemCommon
    attr_accessor :pointer

    def initialize(pointer = 0, delta_time = 0)
      super(SONG_POINTER, delta_time)
      @pointer = pointer
    end

    def data_as_bytes
      data = []
      data << @status
      data << ((@pointer >> 8) & 0xff)
      data << (@pointer & 0xff)
    end

    def to_s
      super << "song ptr #{number_to_s(@pointer)}"
    end
  end

  class SongSelect < SystemCommon
    attr_accessor :song

    def initialize(song = 0, delta_time = 0)
      super(SONG_SELECT, delta_time)
      @song = song
    end

    def data_as_bytes
      data = []
      data << @status
      data << @song
    end

    def to_s
      super << "song sel #{number_to_s(@song)}"
    end
  end

  class TuneRequest < SystemCommon
    def initialize(delta_time = 0)
      super(TUNE_REQUEST, delta_time)
    end

    def data_as_bytes
      data = []
      data << @status
    end

    def to_s
      super << 'tune req'
    end
  end

  class Realtime < Event
    def initialize(status, delta_time)
      super(status, delta_time)
    end

    def data_as_bytes
      data = []
      data << @status
    end

    def to_s
      super << "realtime #{number_to_s(@status)}"
    end
  end

  class Clock < Realtime
    def initialize(delta_time = 0)
      super(CLOCK, delta_time)
    end

    def to_s
      super << 'clock'
    end
  end

  class Start < Realtime
    def initialize(delta_time = 0)
      super(START, delta_time)
    end

    def to_s
      super << 'start'
    end
  end

  class Continue < Realtime
    def initialize(delta_time = 0)
      super(CONTINUE, delta_time)
    end

    def to_s
      super << 'continue'
    end
  end

  class Stop < Realtime
    def initialize(delta_time = 0)
      super(STOP, delta_time)
    end

    def to_s
      super << 'stop'
    end
  end

  class ActiveSense < Realtime
    def initialize(delta_time = 0)
      super(ACTIVE_SENSE, delta_time)
    end

    def to_s
      super << 'act sens'
    end
  end

  class SystemReset < Realtime
    def initialize(delta_time = 0)
      super(SYSTEM_RESET, delta_time)
    end

    def to_s
      super << 'sys reset'
    end
  end

  class MetaEvent < Event
    attr_reader :meta_type, :data

    def self.bytes_as_str(bytes)
      bytes ? bytes.collect { |byte| byte.chr }.join : nil
    end

    def self.str_as_bytes(str)
      str.split(//).collect { |chr| chr.ord }
    end

    def initialize(meta_type, data = nil, delta_time = 0)
      super(META_EVENT, delta_time)
      @meta_type = meta_type
      self.data = (data)
    end

    def data_as_bytes
      data = []
      data << @status
      data << @meta_type
      data << (@data ? Utils.as_var_len(@data.length) : 0)
      data << @data if @data
      data.flatten
    end

    def data_as_str
      MetaEvent.bytes_as_str(@data)
    end

    # Stores bytes. If data is a string, splits it into an array of bytes.
    def data=(data)
      @data = case data
              when String
                MetaEvent.str_as_bytes(data)
              else
                data
              end
    end

    def to_s
      str = super()
      str << "meta #{number_to_s(@meta_type)} "
      # I know, I know...this isn't OO.
      str << case @meta_type
             when META_SEQ_NUM
               'sequence number'
             when META_TEXT
               "text: #{data_as_str}"
             when META_COPYRIGHT
               "copyright: #{data_as_str}"
             when META_SEQ_NAME
               "sequence or track name: #{data_as_str}"
             when META_INSTRUMENT
               "instrument name: #{data_as_str}"
             when META_LYRIC
               "lyric: #{data_as_str}"
             when META_MARKER
               "marker: #{data_as_str}"
             when META_CUE
               "cue point: #{@data}"
             when META_TRACK_END
               'track end'
             when META_SMPTE
               'smpte'
             when META_TIME_SIG
               'time signature'
             when META_KEY_SIG
               'key signature'
             when META_SEQ_SPECIF
               'sequence specific'
             else
               # Some other possible @meta_type values are handled by subclasses.
               '(other)'
             end
      str
    end
  end

  class Marker < MetaEvent
    def initialize(msg, delta_time = 0)
      super(META_MARKER, msg, delta_time)
    end
  end

  class Tempo < MetaEvent
    MICROSECS_PER_MINUTE = 1_000_000 * 60

    # Translates beats per minute to microseconds per quarter note (beat).
    def self.bpm_to_mpq(bpm)
      MICROSECS_PER_MINUTE / bpm
    end

    # Translates microseconds per quarter note (beat) to beats per minute.
    def self.mpq_to_bpm(mpq)
      MICROSECS_PER_MINUTE.to_f / mpq.to_f
    end

    def initialize(msecs_per_qnote, delta_time = 0)
      super(META_SET_TEMPO, msecs_per_qnote, delta_time)
    end

    def tempo
      @data
    end

    def tempo=(val)
      @data = val
    end

    def data_as_bytes
      data = []
      data << @status
      data << @meta_type
      data << 3
      data << ((@data >> 16) & 0xff)
      data << ((@data >> 8) & 0xff)
      data << (@data & 0xff)
    end

    def to_s
      "tempo #{@data} msecs per qnote (#{Tempo.mpq_to_bpm(@data)} bpm)"
    end
  end

  # Container for time signature events
  class TimeSig < MetaEvent
    # Constructor
    def initialize(numer, denom, clocks, qnotes, delta_time = 0)
      super(META_TIME_SIG, [numer, denom, clocks, qnotes], delta_time)
    end

    # Returns the complete event as stored in the sequence
    def data_as_bytes
      data = []
      data << @status
      data << @meta_type
      data << 4
      data << @data[0]
      data << @data[1]
      data << @data[2]
      data << @data[3]
    end

    # Calculates the duration (in ticks) for a full measure
    def measure_duration(ppqn)
      (4 * ppqn * @data[0]) / (2**@data[1])
    end

    # Returns the numerator (the top digit) for the time signature
    def numerator
      @data[0]
    end

    # Returns the denominator of the time signature. Use it as a power of 2
    # to get the displayed (lower-part) digit of the time signature.
    def denominator
      @data[1]
    end

    # Returns the metronome tick duration for the time signature. On
    # each quarter note, there's 24 ticks.
    def metronome_ticks
      @data[2]
    end

    # Returns the time signature for the event as a string.
    # Example: "time sig 3/4"
    def to_s
      "time sig #{@data[0]}/#{2**@data[1]}"
    end
  end

  # Container for key signature events
  class KeySig < MetaEvent
    # Constructor
    def initialize(sharpflat, is_minor, delta_time = 0)
      super(META_KEY_SIG, [sharpflat, is_minor], delta_time)
    end

    # Returns the complete event as stored in the sequence
    def data_as_bytes
      data = []
      data << @status
      data << @meta_type
      data << 2
      data << @data[0]
      data << (@data[1] ? 1 : 0)
    end

    # Returns true if it's a minor key, false if major key
    def minor_key?
      @data[1]
    end

    # Returns true if it's a major key, false if minor key
    def major_key?
      !@data[1]
    end

    # Returns the number of sharps/flats in the key sig. Negative for flats.
    def sharpflat
      @data[0] > 7 ? @data[0] - 256 : @data[0]
    end

    # Returns the key signature as a text string.
    # Example: "key sig A flat major"
    def to_s
      if minor_key?
        "key sig #{minorkeys[sharpflat + 7]} minor"
      else
        "key sig #{majorkeys[sharpflat + 7]} major"
      end
    end

    # Returns the key signature as a code.
    # Example: "Ab" for "key sig A flat major"
    def to_code
      if minor_key?
        minorkey_codes[sharpflat + 7]
      else
        majorkey_codes[sharpflat + 7]
      end
    end

    private

    def majorkeys
      @majorkeys ||= ['C flat', 'G flat', 'D flat', 'A flat', 'E flat', 'B flat', 'F',
                      'C', 'G', 'D', 'A', 'E', 'B', 'F#', 'C#']
    end

    def minorkeys
      @minorkeys ||= ['a flat', 'e flat', 'b flat', 'f', 'c', 'g', 'd',
                      'a', 'e', 'b', 'f#', 'c#', 'g#', 'd#', 'a#']
    end

    def majorkey_codes
      @majorkeys_codes ||= ['Cb', 'Gb', 'Db', 'Ab', 'Eb', 'Bb', 'F',
                            'C', 'G', 'D', 'A', 'E', 'B', 'F#', 'C#']
    end

    def minorkey_codes
      @minorkeys_codes ||= ['Abm', 'Ebm', 'Bbm', 'Fm', 'Cm', 'Gm', 'Dm',
                            'Am', 'Em', 'Bm', 'F#m', 'C#m', 'G#m', 'D#m', 'A#m']
    end
  end
end
